components/esb-tools/plugins/org.wso2.integrationstudio.gmf.esb.diagram/src/org/wso2/integrationstudio/gmf/esb/diagram/part/EsbMultiPageEditor.java

```diff
@@ -1093,19 +1094,54 @@ private IFile updateAssociatedXMLFile(IProgressMonitor monitor) throws Exception
             log.warn("Could not get the source");
             return null;
         }
-        InputStream is = new ByteArrayInputStream(source.getBytes());
-        if (xmlFile.exists()) {
-            xmlFile.setContents(is, true, true, monitor);
-        } else {
-            xmlFile.create(is, true, monitor);
-        }
+        
+        
+        /*******************************************
+         * @author isurumaduranga
+         * Fixing CS0193629
+         * Issue - There's a scheduling rule mismatch when trying to save .dmc and .xml files at the same time
+         * Reason - Trying to save .dmc and .xml files in the same job
+         * Explanation - Eclipse jobs API forbids holding two scheduling rules in the same job to avoid dead locks
+         * ( one for .dmc and the other one for .xml )
+         * Fix - Spawning a new job to update .xml file
+         ********************************************/
+        class UpdateXMLJob extends Job {
+            IFile xmlFile;
+            String source;
+            public UpdateXMLJob(IFile xmlFile, String source) {
+               super("Update Associated XML Job");
+               this.xmlFile = xmlFile;
+               this.source = source;
+            }
+            public IStatus run(IProgressMonitor monitor) {
+                try {
+                    InputStream is = new ByteArrayInputStream(source.getBytes());
+                    if (xmlFile.exists()) {
+                        xmlFile.setContents(is, true, true, monitor);
+                    } else {
+                        xmlFile.create(is, true, monitor);
+                    }
 
-        artifactXMLFiles.add(xmlFile);
-        xmlFile.deleteMarkers(SOURCE_VIEW_ERROR, false, 1);
-        if (getActivePage() == SOURCE_VIEW_PAGE_INDEX && file.findMarkers(CONFIG_ERROR, false, 1).length > 0) {
-            xmlFile.createMarker(SOURCE_VIEW_ERROR);
-        }
+                    artifactXMLFiles.add(xmlFile);
+                    xmlFile.deleteMarkers(SOURCE_VIEW_ERROR, false, 1);
+                    if (getActivePage() == SOURCE_VIEW_PAGE_INDEX && file.findMarkers(CONFIG_ERROR, false, 1).length > 0) {
+                        xmlFile.createMarker(SOURCE_VIEW_ERROR);
+                    }
+                } catch (CoreException e) {
+                    e.printStackTrace();
```

Comment:
Shall we handle this exception properly? maybe we can log using logger

Reply:
Fixed